### ISSUE TITLE
OpenLogic 7.2 - Agent 2.0.18 - ERROR:unicode argument without an encoding occured creating /var/lib/waagent/CustomData

### ISSUE THREAD

**Issue by tasquith** [Azure/WALinuxAgent#187]
When using an image based on the OpenLogic 7.2 image, with an updated waagent via yum, I get the following error when attempted to use custom data.

ERROR:unicode argument without an encoding occured creating /var/lib/waagent/CustomData

the base64 custom data is:

<CustomData>IyEvYmluL2Jhc2gKc2V0IC1lCgpob3N0bmFtZSAiYXAtc2VydmljZS1kaXNjb3Zlcnktc2VydmVyLTAiCmlwdGFibGVzIC1GCnNldGVuZm9yY2UgMApQUklWQVRFX0lQPWBpZmNvbmZpZyB8IGdyZXAgJ2luZXQgYWRkcicgfCBjdXQgLWQgJzonIC1mIDIgfCBhd2sgJ3sgcHJpbnQgJDEgfScgfCBncmVwIC1FICdeKDEwXC4pJ2AKZWNobyAvdXNyL2xvY2FsL2Jpbi9ib290c3RyYXAtY29uc3VsLnNoIHRydWUgJFBSSVZBVEVfSVAgdHJ1ZSAzICRQUklWQVRFX0lQICIiID4+IC90bXAvdHgudHh0Ci91c3IvbG9jYWwvYmluL2Jvb3RzdHJhcC1jb25zdWwuc2ggdHJ1ZSAkUFJJVkFURV9JUCB0cnVlIDMgJFBSSVZBVEVfSVAgIiIKCmV4aXQgMA==</CustomData>

Please find the error logs below:

```
2016/05/10 15:11:17 Azure Linux Agent Version: WALinuxAgent-2.0.18
2016/05/10 15:11:17 Linux Distribution Detected      : centos
2016/05/10 15:11:17 Module /lib/modules/3.10.0-327.10.1.el7.x86_64/kernel/drivers/ata/ata_piix.ko driver for ATAPI CD-ROM is already present.
2016/05/10 15:11:18 mount: /dev/sr0 is write-protected, mounting read-only
2016/05/10 15:11:18 mount: /dev/sr0 mounted on /mnt/cdrom/secure.
2016/05/10 15:11:18 mount succeeded on attempt #1
2016/05/10 15:11:18 VMM Init script not found.  Provisioning for Azure
2016/05/10 15:11:18 IPv4 address: 10.0.35.4
2016/05/10 15:11:18 MAC  address: 00:0D:3A:B2:F2:B4
2016/05/10 15:11:18 Probing for Azure environment.
2016/05/10 15:11:18 DoDhcpWork: Setting socket.timeout=10, entering recv
2016/05/10 15:11:18 Set default gateway: 10.0.35.1
2016/05/10 15:11:18 add default gateway result: 0
2016/05/10 15:11:18 Discovered Azure endpoint: 168.63.129.16
2016/05/10 15:11:28 ERROR:Socket IOError timed out, args:('timed out',)
2016/05/10 15:11:28 ERROR:Retry=0
2016/05/10 15:11:28 ERROR:HTTP Req: GET /?comp=versions
2016/05/10 15:11:28 ERROR:HTTP Req: Data=None
2016/05/10 15:11:28 ERROR:HTTP Req: Header=None
2016/05/10 15:11:28 ERROR:HTTP Err: response is empty.
2016/05/10 15:11:38 Fabric preferred wire protocol version: 2015-04-05
2016/05/10 15:11:38 Negotiated wire protocol version: 2012-11-30
2016/05/10 15:11:38 SetBlockDeviceTimeout: Update the device sda with timeout 300
2016/05/10 15:11:38 SetBlockDeviceTimeout: Update the device sdb with timeout 300
2016/05/10 15:11:39 OS.UpdateRdmaDriver configured to False so skip the rdma update.
2016/05/10 15:11:39 OS.CheckRdmaDriver configured to False so skip the rdma check.
2016/05/10 15:11:39 Retrieved GoalState from Azure Fabric.
2016/05/10 15:11:39 ExpectedState: Started
2016/05/10 15:11:39 ContainerId: dc58cd60-09dd-4e60-a9f0-be8e7f0d64ee
2016/05/10 15:11:39 RoleInstanceId: 24b78327-a979-4809-88db-3c79818e7a03._ap-service-discovery-server-0
2016/05/10 15:11:40 Public cert with thumbprint: <<REDACTED>> was retrieved.
2016/05/10 15:11:41 Provisioning image started.
2016/05/10 15:11:41 Detect GPT...
2016/05/10 15:11:43 mount: /dev/sr0 is write-protected, mounting read-only
2016/05/10 15:11:43 mount succeeded on attempt #1
2016/05/10 15:11:43 Provisioning image using OVF settings in the DVD.
2016/05/10 15:11:43 ERROR:unicode argument without an encoding occured creating /var/lib/waagent/CustomData
2016/05/10 15:11:43 Disabled SSH password-based authentication methods.
2016/05/10 15:11:44 EnvMonitor: Detected host name change: localhost.localdomain -> ap-service-discovery-server-0
2016/05/10 15:11:44 Setting host name: ap-service-discovery-server-0
2016/05/10 15:11:44 Created user account: centos
2016/05/10 15:11:44 Deploy public key: <<REDACTED>>
2016/05/10 15:11:46 Set default gateway: 10.0.35.1
2016/05/10 15:11:46 add default gateway result: 0
2016/05/10 15:11:47 Ovf XML process finished
2016/05/10 15:11:48 Posted Role Properties. CertificateThumbprint=<<REDACTED>>
2016/05/10 15:11:48 Root password deleted.
2016/05/10 15:11:48 Provisioning image completed.
2016/05/10 15:11:48 Posted Role Properties. CertificateThumbprint=<<REDACTED>>
2016/05/10 15:11:48 Resource disk (/dev/sdb1) is mounted at /mnt/resource with fstype ext4
2016/05/10 15:11:52 Finished processing ExtensionsConfig.xml
2016/05/10 15:11:52 Successfully reported handler status
```

**Comment by szarkos** [comment]
This issue was fixed with this PR: https://github.com/Azure/WALinuxAgent/commit/7e2484d57bba82a24f2f8c4bbb5b83f83b473940.  Waiting on a new agent release to resolve the issue in the package.

The base64-encoded data is still available in /var/lib/waagent/ovf-env.xml.  Is this a major blocker for you?

**Comment by tomasquith** [comment]
Thanks @szarkos, 

Yes this is stopping us bootstrapping nodes as part of a terraform workflow on azure, holding up the port of our platform from AWS. 

When is the next agent release planned?

Tom

**Comment by hglkrijger** [comment]
I took a look at this issue, and the same bug should not exist in the 2.1 agent, which uses utf-8 encoding by default. We will be updating the agent to 2.1.3 in the near future, though in the mean time you could get it from here: https://github.com/Azure/WALinuxAgent/releases/tag/v2.1.3 

@tomasquith, does this help unblock you?

**Comment by ahmetalpbalkan** [comment]
also, here's the instructions on how to update the agent: https://azure.microsoft.com/documentation/articles/virtual-machines-linux-update-agent/

**Comment by hglkrijger** [comment]
fix merged into develop

**Comment by tomasquith** [comment]
Thanks for your help @hglkrijger and @ahmetalpbalkan 

Have you got any idea as to when this agent release will be pushed to the repos?

**Comment by hglkrijger** [comment]
@tomasquith, we are starting validation this week, and hope to have it out soon.

**Comment by tasquith** [comment]
@hglkrijger Thanks Hans :)
